File: workers/python/src/pilar4/metrics.py
"""Métricas do funil e ARR projetado (Pilar 4 — Python)."""
from __future__ import annotations


def _pct(a: int, b: int) -> float:
    return (a / b * 100.0) if b > 0 else 0.0


def funnel_metrics(c: dict) -> dict:
    """c: sent, delivered, read, replied, qualified, scheduled, closed, rejected."""
    return {
        "delivery_rate": round(_pct(c.get("delivered", 0), c.get("sent", 0)), 1),
        "read_rate": round(_pct(c.get("read", 0), c.get("delivered", 0)), 1),
        "reply_rate": round(_pct(c.get("replied", 0), c.get("read", 0)), 1),
        "qualification_rate": round(_pct(c.get("qualified", 0), c.get("replied", 0)), 1),
        "schedule_rate": round(_pct(c.get("scheduled", 0), c.get("qualified", 0)), 1),
        "close_rate": round(_pct(c.get("closed", 0), c.get("scheduled", 0)), 1),
    }


def arr_projected(qualified: int, schedule_rate: float, close_rate: float, ticket_monthly: float) -> float:
    clients = qualified * (schedule_rate / 100.0) * (close_rate / 100.0)
    return round(clients * ticket_monthly * 12, 2)

File: packages/core/src/pilar4/metrics.ts
/** Métricas do funil e ARR projetado (Pilar 4). */
import type { Channel } from "../pilar2/types.js";

export interface FunnelCounters {
  sent: number;
  delivered: number;
  read: number;
  replied: number;
  qualified: number;
  scheduled: number;
  closed: number;
  rejected: number;
}

export interface FunnelMetrics {
  deliveryRate: number;
  readRate: number;
  replyRate: number;
  qualificationRate: number;
  scheduleRate: number;
  closeRate: number;
}

const pct = (a: number, b: number): number => (b > 0 ? (a / b) * 100 : 0);

export function funnelMetrics(c: FunnelCounters): FunnelMetrics {
  return {
    deliveryRate: pct(c.delivered, c.sent),
    readRate: pct(c.read, c.delivered),
    replyRate: pct(c.replied, c.read),
    qualificationRate: pct(c.qualified, c.replied),
    scheduleRate: pct(c.scheduled, c.qualified),
    closeRate: pct(c.closed, c.scheduled),
  };
}

/** ARR projetado: clientes esperados × ticket × 12 (com probabilidade de fechamento). */
export function arrProjected(qualified: number, scheduleRate: number, closeRate: number, ticketMonthly: number): number {
  const clients = qualified * (scheduleRate / 100) * (closeRate / 100);
  return clients * ticketMonthly * 12;
}

export interface ChannelCounters {
  channel: Channel;
  sent: number;
  delivered: number;
  read: number;
  replied: number;
  rejected: number;
}

export function channelRate(c: ChannelCounters): FunnelMetrics {
  return funnelMetrics({
    sent: c.sent,
    delivered: c.delivered,
    read: c.read,
    replied: c.replied,
    qualified: 0,
    scheduled: 0,
    closed: 0,
    rejected: c.rejected,
  });
}
